src/metabase/db/util.clj

```diff
@@ -66,3 +68,71 @@
   ;;    [:in :semantic_type #{"type/URL" "type/ImageURL"}]
   ([expr type-keyword]
    [:in expr (type-keyword->descendants type-keyword)]))
+
+(defmacro with-conflict-retry
+  "Retry a database mutation a single time if it fails due to concurrent insertions.
+   May retry for other reasons."
+  [& body]
+  `(try
+     ~@body
+     (catch ExceptionInfo e#
+       ;; The underlying exception thrown by the driver is database specific and opaque, so we treat any exception as a
+       ;; possible database conflict due to a concurrent insert. If we want to be more conservative, we would need
+       ;; a per-driver or driver agnostic way to test the exception.
+       ~@body)))
+
+(defn select-or-insert!
+  "Return a database record if it exists, otherwise create it.
+
+   This is more general than using `UPSERT`, `MERGE` or `INSERT .. ON CONFLICT`, and it also allows one to avoid
+   calculating initial values that may be expensive, or require side effects.
+
+   In the case where there is an underlying db constraint to prevent duplicates, this method takes care of handling
+   rejection from the database due to a concurrent insert, and will retry a single time to pick up the existing row.
+   This may result in `insert-fn` being called a second time.
+
+   In the case where there is no underlying db constraint, concurrent calls may still result in duplicates.
+   To prevent this in a database agnostic way, during an existing non-serializable transaction, would be non-trivial."
+  {:style/indent 2}
+  [model select-map insert-fn]
+  (with-conflict-retry
+   (let [select-kvs (mapcat identity select-map)
+         validate   (fn [updated#]
+                      (assert (= select-map (merge select-map (select-keys updated# (keys select-map))))
+                              "This should not be used to change any of the identifying values")
+                      ;; For convenience, we allow the insert-fn to omit fields in the search-map
+                      (merge updated# select-map))]
+     (or (apply t2/select-one model select-kvs)
+         (t2/insert-returning-instance! model (validate (insert-fn)))))))
+
+(defn update-or-insert!
+  "Update a database record, if it exists, otherwise create it.
+
+   This is more general than using `UPSERT`, `MERGE` or `INSERT .. ON CONFLICT`, and it also allows one to avoid
+   calculating initial values that may be expensive, or require side effects.
+
+   In the case where there is an underlying db constraint to prevent duplicates, this method takes care of handling
+   rejection from the database due to a concurrent insert, and will retry a single time to pick up the existing row.
+   This may result in `insert-fn` being called a second time.
+
+   In the case where there is no underlying db constraint, concurrent calls may still result in duplicates.
+   To prevent this in a database agnostic way, during an existing non-serializable transaction, would be non-trivial."
+  {:style/indent 2}
```

Comment:
Why? I don't remember seeing many functions with custom indents so this is surprising

Reply:
Another macro remnant